api/instance.ts

```diff
@@ -1,10 +1,50 @@
 import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
 
+interface ApiInstanceOptions {
+  axiosConfig: AxiosRequestConfig;
+  fallbackUrl?: string;
+}
+
 export default class ApiInstance {
   bitcoinApi: AxiosInstance;
 
-  constructor(config: AxiosRequestConfig) {
-    this.bitcoinApi = axios.create(config);
+  fallbackBitcoinApi?: AxiosInstance;
+
+  constructor(options: ApiInstanceOptions) {
+    const { axiosConfig, fallbackUrl } = options;
+
+    this.bitcoinApi = axios.create(axiosConfig);
+
+    if (fallbackUrl) {
+      this.fallbackBitcoinApi = axios.create({ ...axiosConfig, baseURL: fallbackUrl });
+      this.bitcoinApi.interceptors.response.use(
+        // if the request succeeds, we do nothing.
+        (response) => response,
+        (error) => {
+          if (!this.fallbackBitcoinApi) {
+            return Promise.reject(error);
+          }
+
+          // if the request times out, we retry on the fallbackBitcoinApi
+          if (error?.code === 'ECONNABORTED') {
+            return this.fallbackBitcoinApi.request({
+              ...error.config,
+              baseURL: fallbackUrl,
+            });
+          }
+
+          // if an address has > 500 UTXOs, mempool.space returns a 400 error,
+          // so we retry on the fallbackBitcoinApi
+          if (error?.response?.status >= 400) {
+            return this.fallbackBitcoinApi.request({
+              ...error.config,
+              baseURL: fallbackUrl,
+            });
+          }
```

Comment:
This class is becoming very tied to the Esplora implementation and nothing else is actually using it. Could we rather delete this file and rather move the stuff it implements directly onto the esplora api provider?

Reply:
absolutely can do that!

Reply:
@victorkirov done